File: backend/app/routers/categories.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import Optional

from app.database import get_db
from app.dependencies import get_current_user
from app.models import Category
from app.schemas.category import CategoryCreate, CategoryUpdate, CategoryOut

router = APIRouter(prefix="/api/categories", tags=["分类"], dependencies=[Depends(get_current_user)])


@router.get("", response_model=list[CategoryOut])
def list_categories(db: Session = Depends(get_db)):
    return db.query(Category).order_by(Category.sort_order).all()


@router.post("", response_model=CategoryOut, status_code=status.HTTP_201_CREATED)
def create_category(body: CategoryCreate, db: Session = Depends(get_db)):
    cat = Category(**body.model_dump())
    db.add(cat)
    db.commit()
    db.refresh(cat)
    return cat


@router.put("/{category_id}", response_model=CategoryOut)
def update_category(category_id: int, body: CategoryUpdate, db: Session = Depends(get_db)):
    cat = db.query(Category).filter(Category.id == category_id).first()
    if not cat:
        raise HTTPException(status_code=404, detail="分类不存在")
    for key, value in body.model_dump(exclude_unset=True).items():
        setattr(cat, key, value)
    db.commit()
    db.refresh(cat)
    return cat


@router.delete("/{category_id}")
def delete_category(category_id: int, db: Session = Depends(get_db)):
    cat = db.query(Category).filter(Category.id == category_id).first()
    if not cat:
        raise HTTPException(status_code=404, detail="分类不存在")
    if cat.subscriptions:
        raise HTTPException(status_code=400, detail="该分类下仍有订阅，无法删除")
    db.delete(cat)
    db.commit()
    return {"detail": "分类已删除"}

File: frontend/src/api/categories.ts
import api from '../composables/useApi'

export function listCategories() {
  return api.get('/categories')
}

export function createCategory(data: { name: string; icon?: string | null; color?: string | null; sort_order?: number }) {
  return api.post('/categories', data)
}

export function updateCategory(id: number, data: { name?: string; icon?: string | null; color?: string | null; sort_order?: number }) {
  return api.put(`/categories/${id}`, data)
}

export function deleteCategory(id: number) {
  return api.delete(`/categories/${id}`)
}

export function reorderCategories(ids: number[]) {
  return api.post('/categories/reorder', { ids })
}